### ISSUE TITLE
Council file method

### ISSUE THREAD

**Issue by Gr770** [Gr770/CK3-Community-Title-Project#56]
Discussion for use of localizing council positions

1. Original branch:
    Edits the original files 00_council_positions and council_l_language.yml
    Mostly likely better optimized, will need testing on somebody's computer that doesnt have a SSD. Has less new files

2. Alt Branch
    Edits the original file council_l_language.yml only
    Better compatibility as it will have our own files. Probably easier to make sub-mods for other big mods like swmh.

**Comment by Charanraja123** [comment]
@Gr770 You are kind of still in CK2 terms, In CK3 you can override terms within a vanilla game file without overriding the game file.
You do not need to edit/include the original file wholly, Check my localization, I only have terms from the original that I needed to overwrite council_l_language.yml file.

**Comment by Charanraja123** [comment]
Quoted this[ forum post](https://www.crusaderkings.com/en/news/ck3-dev-diary-37-making-mods-1410656-1234)
Combining multiple mods together has always been a bit tricky but I wanted to try and make some gains with compatibility so that it doesn’t require as many manual patches. 

To aid in this a bit I’ve made it so database entries can be overridden by key even if that new entry is in a different file. This means if you want to override what the lunatic trait does then instead of needing to copy the entire traits file just to change one entry you can just make your own file just containing your new definition of what the lunatic trait does, such as giving it a boost in learning, prowess, and attraction opinion like I have done below!

This also applies to files from other mods, as long as yours is loaded first, so you can overwrite traits and other database elements from other mods

**Comment by Charanraja123** [comment]
Hope this helps you understand.

**Comment by Gr770** [comment]
@Charanraja123 I have already tested it. I could not change the positions without changing the original file.

Either it's some weird priority issue where the original file is overwriting the new file, or you would still have to change that file.

![image](https://user-images.githubusercontent.com/47698688/94372426-2d5a3f80-00b2-11eb-947d-35cd96e7075a.png)

**Comment by Gr770** [comment]
This is from copying and pasting that original and replacing it with the new localization

![image](https://user-images.githubusercontent.com/47698688/94372461-980b7b00-00b2-11eb-8922-ca3b1b148008.png)

**Comment by Charanraja123** [comment]
OMG. I'm really sorry.
I made some stupid mistakes which I wont embarrass myself telling it here. I am ashamed, We can continue on with the discussion.
Sorry again.

**Comment by Charanraja123** [comment]
> 
> 
> Discussion for use of localizing council positions
> 
>     1. Original branch:
>        Edits the original files 00_council_positions and council_l_language.yml
>        Mostly likely better optimized, will need testing on somebody's computer that doesnt have a SSD. Has less new files
> 
>     2. Alt Branch
>        Edits the original file council_l_language.yml only
>        Better compatibility as it will have our own files. Probably easier to make sub-mods for other big mods like swmh.

Regarding performance, I don't expect that much of a difference, especially since I think these localizations are only called in when viewing the council tab, unlike title flavorisations which are applied everywhere (AFAIK). In both of these cases, it should not matter much compared to other parts of our mod.

**Comment by Gr770** [comment]
I dont think its bad either. I actually perfer the alt way.

**Comment by Charanraja123** [comment]
I mean, I obviously prefer alt-way too , especially for the compatibility part. 
I don't know whether anyone else will have an opinion on the coding structure\method based on compatibility or performance metrics. Should we leave this open for discussion or start adding in titles for the alt method?

**Comment by Viridianus** [comment]
I don't think we should judge performance much. Compatibility is more important and more testable for non-specialists, and there's a simple rule for it: the less files you edit, the better. Why would we need to edit the 00_ file anyway?

**Comment by Charanraja123** [comment]
For his method, He called in the base game function directly, so he needed to do that. But as you say, less edited files is the better (not always though). As i said before, the most performance heavy component in our mod, till now, is the title flavorisations, I don't think these compare to those. CK3 is efficient and designed good enough (As of now) for these to not matter.
But then again I use an SSD, maybe it might affect people who use HDDs.

**Comment by Viridianus** [comment]
Yeah, and I don't even want to estimate how much heavier the EtB-based submod will be... So let's imagine everyone's using a supercomputer :D I vote for alt-method.

**Comment by Gr770** [comment]
The only issue lies in the Spouse and Court Chaplin

**Comment by Charanraja123** [comment]
Okay. First I'll address some changes I made and tested (Locally, Not uploaded in Branch)
1) Changed custom calling from `GetPlayer.Custom` to `ROOT.Char.Custom`, So now the Councillor is the root scope instead of the player. I changed this because having Player as the root scope brought out some problems in Leige's council view.
2) Following that changed `has_culture` to `liege.culture`, since the root has changed now.
3) This also should solve the is_male problem as the root is the Councillor and it check for their gender.

**Comment by Charanraja123** [comment]
> The only issue lies in the Spouse and Court Chaplin

I'll explain the base game mechanics here

## Court chaplain
Obligatory rant, The way the vanilla game handles this is so bad, but I'll explain it to the best of my ability. 

- First the game check whether the liege's faith has the temporal doctrine in **_00_council_positions.txt_**

     - If it does, a localisation called **actual_bishop_title** is called in. This is defined in the **_divinity_custom_loc_l_english.yml_** .Why? Because the **actual_bishop_title** calls in a custom localisation of its own called `GetActualBishopTitle` defined in **_00_divinity_custom_loc.txt_**
    - Now going into `GetActualBishopTitle`, If check whether the character is either landed or the head of the faith. If they are landed, it gives them their ruler title. If head of faith, it gives them the head of faith title of their faith. If not both of these, Then it goes on to check faith and `liege.highest_held_title_tier`, assign localization based on it. These titles are actually localized in our very own **_council_l_language.yml_** 

- Now back to our first point, what if the liege's faith doesn't have the temporal doctrine. 
  - If check whether you the are a kingdom/empire following ismali  faith or have the ashari religion (for every tier induvidually) and handles the localization key appropriately. 

- So what about the rest of them? Congrats, you just got default court chaplain.
***
## Spouse
- This is relatively simple, It checks conditions based on highest_held_tier and cultures and assigns localizations for them.

**Comment by Charanraja123** [comment]
Now I'll explain how we can modify these using our own Custom Loc.

## Court Chaplain
-First step would be to changed the **actual_bishop_title** localisation of 'GetActualBishopTitle` into something of our own like `CPTBishopTitle` (or something like that). You might ask, why not just go into `GetActualBishopTitle` and edit things. The problem is compatibility, there is a lot of other stuff in **_00_divinity_custom_loc.txt_** that we will need to overwrite to edit this particular function. Therefore, we are better off creating our own.
- Next question, is do we need to replace the **_divinity_custom_loc_l_english.yml_** to edit **actual_bishop_title**, my answer would be yes considering yesterday's episodes, but it is upto to testing.
- What about ismali and ashari defined localisations, Honestly we could just overwrite them since we are already editing the **_council_l_language.yml_** file
- We could create another one for defaults, which is basically faiths that do not have a temporal doctrine and are not part of either ismaili or ashari faiths. This should be simple as overriding the default court_chaplain like all other council positions we have done.

Note; Since landed titles are an integral part of this, I think we can put off Court Chaplain until we finish theocratic titles.

***
## Spouse
Okay, so this one is funny, because  I think this is the easiest ever. We just need to replace every spouse localisation with `ROOT.Char.GetTitleAsName`. This works because we already have defined spouse flavorisations. Why am i sure it should work? Because this is the same method used by Paradox themselves for landed bishops in the `GetActualBishopTitle`. Why didn't Paradox use this? **I have no fucking Idea**. If it doesn;t work, again we just replace localization with the same custom function. Should not be a problem at all.

TLDR : Custom localization can be used to change all these titles, and is more compatible than the other method even for these.

**Comment by Viridianus** [comment]
I fully accept the premise on spouses (though note that landed spouses are not your councillors, that's what may have stopped Pdx). As for court chaplain: exactly what would the default (non-temporal, non-Ismaili, non-Ashari) court chaplain be?

**Comment by Charanraja123** [comment]
My understanding is that Every spouse regardless of their landed status obtain a title based on rulers except for Mayors, right?

For the court chaplain, from a historical point of view I do not know, Maybe they were not any. But if any exists or are created by the players, it is possible to Flavourise them is what I'm saying. For now even Ismaili duchies or lower do just go to the default title (Court Chaplain), which may not have been the case I presume.

**Comment by Viridianus** [comment]
Oh, you mean the default title is literally "Court Chaplain" for English?

Also, I'm kinda confused about what branch is to be edited if I want to add/change some council localizations. Is Council-Positions-Redo obsolete now?

**Comment by Charanraja123** [comment]
Yeah its literally "court chaplain".

The code is not settled in yet, So I would say just hold on to your changes for a week, at max. I'll try to clean everything up by then.
Most likely that branch is deprecated, but @Gr770 should confirm that

**Comment by Gr770** [comment]
3/5 rule in favor of alt method and we have a clear path to complete the localization without 00_council_positions.txt. 

When I get home I'll back a tag for the old method called 'backup/depreciated just in case we want it again.

**Comment by Charanraja123** [comment]
Don't we have six contributors? xD

**Comment by Gr770** [comment]
The czech guy was never an official contributor. He submitted a request to fix the czech culture name.

**Comment by Charanraja123** [comment]
I've finished chancellors, Please test it and let me know if it works as intended.

**Comment by Charanraja123** [comment]
Can we close this? Since we finished the discussion?

**Comment by Viridianus** [comment]
I thought we'd leave this just in case until the merge is performed, and the merge will close this automatically. But, so that it's not a deadweight, I'll ask here about Council-Positions-Redo _again_. Maybe @Gr770 should delete or hide that branch?

**Comment by Charanraja123** [comment]
He said he'll deprecate it or something. I'm sure he'll do it today or tomorrow. Maybe he's waiting for me to finish up and then test stuff.
